crates/test-framework/src/conformance.rs

```diff
@@ -0,0 +1,699 @@
+use proc_macro2::{Ident, TokenStream};
+
+fn equalf64() -> TokenStream {
+    quote! {
+      fn equal(x: f64, y: f64, max_ulp_diff: usize) -> bool {
+        if x.is_nan() != y.is_nan() {  // one is nan but the other is not
+            return false;
+        }
+        if x.is_nan() && y.is_nan() {
+            return true;
+        }
+        if x.is_infinite() != y.is_infinite() {  // one is inf but the other is not
+            return false;
+        }
+        if x.is_infinite() != y.is_infinite() {  // one is inf but the other is not
+            return false;
+        }
+        let xi: i64 = unsafe { core::intrinsics::transmute(x) };
+        let yi: i64 = unsafe { core::intrinsics::transmute(y) };
+        if (xi < 0) != (yi < 0) { // different sign
+          return false;
+        }
+        let ulps = (xi - yi).abs() as usize;
+        ulps <= max_ulp_diff
+      }
+    }
+}
+
+fn equalf32() -> TokenStream {
+    quote! {
+      fn equal(x: f32, y: f32, max_ulp_diff: usize) -> bool {
+        if x.is_nan() != y.is_nan() {  // one is nan but the other is not
+            return false;
+        }
+        if x.is_nan() && y.is_nan() {
+            return true;
+        }
+        if x.is_infinite() != y.is_infinite() {  // one is inf but the other is not
+            return false;
+        }
+        let xi: i32 = unsafe { core::intrinsics::transmute(x) };
+        let yi: i32 = unsafe { core::intrinsics::transmute(y) };
+        if (xi < 0) != (yi < 0) { // different sign
+          return false;
+        }
+        let ulps = (xi - yi).abs() as usize;
+        ulps <= max_ulp_diff
+      }
+    }
+}
+
+fn equali32() -> TokenStream {
+    quote! {
+      fn equal(x: i32, y: i32, max_ulp_diff: usize) -> bool {
+        let ulps = (x - y).abs() as usize;
+        ulps <= max_ulp_diff
+      }
+    }
+}
+
+macro_rules! forward {
+  ($func:ident, $x:ident) => (
+      pub fn $func(ulp: usize) -> TokenStream  {
+        let equal: TokenStream = paste::expr!{ [<equal $x>]() };
+        quote! {
+          #[cfg(test)]
+          mod conformance {
```

Comment:
This is also expanded many times. Do you think it could be possible to move this into the test framework, and use generic parameters to avoid expanding this code for all functions ?